src/spreadinterp.cpp

```diff
@@ -2047,254 +1610,658 @@ void get_subgrid(BIGINT &offset1, BIGINT &offset2, BIGINT &offset3, BIGINT &padd
+                        T *FINUFFT_RESTRICT kx, T *FINUFFT_RESTRICT ky,
+                        T *FINUFFT_RESTRICT kz, const T *data_nonuniform,
+                        const finufft_spread_opts &opts, int did_sort)
+// Spread NU pts in sorted order to a uniform grid. See spreadinterp() for doc.
+{
+  CNTime timer{};
+  const auto ndims = ndims_from_Ns(N1, N2, N3);
+  const auto N     = N1 * N2 * N3;             // output array size
+  const auto ns    = opts.nspread;             // abbrev. for w, kernel width
+  auto nthr        = MY_OMP_GET_MAX_THREADS(); // guess # threads to use to spread
+  if (opts.nthreads > 0) nthr = opts.nthreads; // user override, now without limit
+#ifndef _OPENMP
+  nthr = 1;                                    // single-threaded lib must override user
+#endif
+  if (opts.debug)
+    printf("\tspread %dD (M=%lld; N1=%lld,N2=%lld,N3=%lld), nthr=%d\n", ndims,
+           (long long)M, (long long)N1, (long long)N2, (long long)N3, nthr);
+  timer.start();
+  std::fill(data_uniform, data_uniform + 2 * N, 0.0); // zero the output array
+  if (opts.debug) printf("\tzero output array\t%.3g s\n", timer.elapsedsec());
+  if (M == 0)                                         // no NU pts, we're done
+    return 0;
 
-template<uint8_t ns, uint8_t kerevalmeth, class T, class simd_type, typename... V>
-auto ker_eval(FLT *FINUFFT_RESTRICT ker, const finufft_spread_opts &opts,
-              const V... elems) noexcept {
-  /* Utility function that allows to move the kernel evaluation outside the spreader for
-     clarity
-     Inputs are:
-     ns = kernel width
-     kerevalmeth = kernel evaluation method
-     T = (single or double precision) type of the kernel
-     simd_type = xsimd::batch for Horner
-     vectorization (default is the optimal simd size)
-     finufft_spread_opts as Horner needs
-     the oversampling factor
-     elems = kernel arguments
-     Examples usage is
-     ker_eval<ns,kerevalmeth>(opts, x, y, z) // for 3D or
-     ker_eval<ns, kerevalmeth>(opts, x, y) // for 2D or
-     ker_eval<ns, kerevalmeth>(opts, x) // for 1D
-   */
-  const std::array inputs{elems...};
-  // compile time loop, no performance overhead
-  for (auto i = 0; i < sizeof...(elems); ++i) {
-    // compile time branch no performance overhead
-    if constexpr (kerevalmeth == 1) {
-      if (opts.upsampfac == 2.0) {
-        eval_kernel_vec_Horner<ns, 200, simd_type>(ker + (i * MAX_NSPREAD), inputs[i],
-                                                   opts);
-      }
-      if (opts.upsampfac == 1.25) {
-        eval_kernel_vec_Horner<ns, 125, simd_type>(ker + (i * MAX_NSPREAD), inputs[i],
-                                                   opts);
-      }
+  auto spread_single = (nthr == 1) || (M * 100 < N); // low-density heuristic?
+  spread_single      = false;                        // for now
+  timer.start();
+  if (spread_single) { // ------- Basic single-core t1 spreading ------
```

Comment:
at some point I need help figuring out to remove this logic - for a future discussion :)